src/object/srv_obj_migrate.c

```diff
@@ -0,0 +1,1729 @@
+	}
+
+	/* Check if the pool tls exists */
+	pool_tls = migrate_pool_tls_lookup_create(pool, migrate_in->om_version,
+						  po_hdl_uuid, co_hdl_uuid,
+						  migrate_in->om_max_eph);
+	if (pool_tls == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	/* NB: only create this tree on xstream 0 */
+	rc = migrate_tree_get_hdl(pool_tls, &btr_hdl);
+	if (rc)
+		D_GOTO(out, rc);
+
+	/* Insert these oids/conts into the local tree */
+	for (i = 0; i < oids_count; i++) {
+		/* firstly insert/check rebuilt tree */
+		rc = migrate_obj_insert(btr_hdl, co_uuid, oids[i], ephs[i],
+					shards[i], migrate_in->om_tgt_idx);
+		if (rc == -DER_EXIST) {
+			D_DEBUG(DB_TRACE, DF_UOID"/"DF_UUID"exists.\n",
+				DP_UOID(oids[i]), DP_UUID(co_uuid));
+			rc = 0;
+			continue;
+		} else if (rc < 0) {
+			D_ERROR("insert "DF_UOID"/"DF_U64" "DF_UUID
+				" shard %u to rebuilt tree failed, rc %d.\n",
+				DP_UOID(oids[i]), ephs[i],
+				DP_UUID(co_uuid), shards[i], rc);
+			break;
+		}
+	}
+	if (rc < 0)
+		D_GOTO(out, rc);
+
+	/* Check and create task to iterate the to-be-rebuilt tree */
+	if (!pool_tls->mpt_ult_running) {
+		pool_tls->mpt_ult_running = 1;
+		migrate_pool_tls_get(pool_tls);
+		rc = dss_ult_create(migrate_ult, pool_tls, DSS_ULT_REBUILD,
+				    DSS_TGT_SELF, 0, NULL);
+		if (rc) {
+			pool_tls->mpt_ult_running = 0;
+			migrate_pool_tls_put(pool_tls);
+			D_ERROR("Create migrate ULT failed: rc %d\n", rc);
+		}
+	}
+out:
+	if (pool)
+		ds_pool_put(pool);
+
+	if (pool_tls)
+		migrate_pool_tls_put(pool_tls);
+	migrate_out = crt_reply_get(rpc);
+	migrate_out->om_status = rc;
+	dss_rpc_reply(rpc, DAOS_REBUILD_DROP_OBJ);
+}
+
+struct migrate_query_arg {
+	uuid_t	pool_uuid;
+	ABT_mutex status_lock;
+	uint64_t migrate_obj_count;
+	uint64_t migrate_rec_count;
+	uint64_t migrate_total_size;
+	uint32_t obj_generated_ult;
+	uint32_t obj_executed_ult;
+	uint32_t generated_ult;
+	uint32_t executed_ult;
+	uint32_t version;
+	int	 status;
+};
+
+static int
+migrate_check_one(void *data)
+{
+	struct migrate_query_arg	*arg = data;
+	struct migrate_pool_tls		*tls;
+
+	tls = migrate_pool_tls_lookup(arg->pool_uuid, arg->version);
+	if (tls == NULL)
+		return 0;
+
+	ABT_mutex_lock(arg->status_lock);
+	arg->migrate_rec_count += tls->mpt_rec_count;
+	arg->migrate_obj_count += tls->mpt_obj_count;
+	arg->migrate_total_size += tls->mpt_size;
+	arg->obj_generated_ult += tls->mpt_obj_generated_ult;
+	arg->obj_executed_ult += tls->mpt_obj_executed_ult;
+	arg->generated_ult += tls->mpt_generated_ult;
+	arg->executed_ult += tls->mpt_executed_ult;
+	if (arg->status)
+		arg->status = tls->mpt_status;
+
+	ABT_mutex_unlock(arg->status_lock);
+	return 0;
+}
+
+int
+ds_migrate_query_status(uuid_t pool_uuid, uint32_t ver,
+			struct ds_migrate_status *dms)
+{
+	struct migrate_query_arg	arg = { 0 };
+	int				rc;
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	arg.version = ver;
+	ABT_mutex_create(&arg.status_lock);
+
+	rc = dss_thread_collective(migrate_check_one, &arg, 0, DSS_ULT_REBUILD);
```

Comment:
hmm, this is to collect the status from all xstreams, and add them together.  Since the status will be shared by all xstream. 
I am not sure you can remove the lock by that reduce func, since the reduce callback from every xstream.